src/nodes/NodeManager.ts

```diff
@@ -573,169 +641,153 @@ class NodeManager {
+        case 'scheduled':
+          {
+            // If it's scheduled then reset delay
+            existingTasks[bucketIndex] = true;
+            // Total delay is refreshBucketDelay + time since task creation
+            const spread =
+              (Math.random() - 0.5) *
+              this.refreshBucketDelay *
+              this.refreshBucketDelaySpread;
+            const delay =
+              performance.now() +
+              performance.timeOrigin -
+              task.created.getTime() +
+              this.refreshBucketDelay +
+              spread;
+            await this.taskManager.updateTask(task.id, { delay }, tran);
+          }
+          break;
+        case 'queued':
+        case 'active':
+          // If it's running then leave it
+          existingTasks[bucketIndex] = true;
+          break;
+        default:
+          // Otherwise, ignore it, should be re-created
+          existingTasks[bucketIndex] = false;
+      }
+      this.logger.info('Set up refreshBucket tasks');
+    }
 
+    // 2. Recreate any missing tasks for buckets
     for (
       let bucketIndex = 0;
-      bucketIndex < this.nodeGraph.nodeIdBits;
+      bucketIndex < existingTasks.length;
       bucketIndex++
     ) {
-      const deadline = Date.now() + this.refreshBucketTimerDefault;
-      this.refreshBucketDeadlineMap.set(bucketIndex, deadline);
+      const exists = existingTasks[bucketIndex];
+      if (!exists) {
+        // Create a new task
+        this.logger.debug(
+          `Creating refreshBucket task for bucket ${bucketIndex}`,
+        );
+        const spread =
+          (Math.random() - 0.5) *
+          this.refreshBucketDelay *
+          this.refreshBucketDelaySpread;
+        await this.taskManager.scheduleTask({
+          handlerId: this.refreshBucketHandlerId,
+          delay: this.refreshBucketDelay + spread,
+          lazy: true,
+          parameters: [bucketIndex],
+          path: ['refreshBucket', `${bucketIndex}`],
+          priority: 0,
+        });
+      }
     }
+    this.logger.info('Set up refreshBucket tasks');
   }
 
-  private async stopRefreshBucketTimers() {
-    clearTimeout(this.refreshBucketTimer);
-  }
-
-  private refreshBucketTimerReset(timeout: number) {
-    clearTimeout(this.refreshBucketTimer);
-    this.refreshBucketTimer = setTimeout(() => {
-      this.refreshBucketRefreshTimer();
-    }, timeout);
-  }
-
-  public refreshBucketUpdateDeadline(bucketIndex: NodeBucketIndex) {
-    // Update the map deadline
-    this.refreshBucketDeadlineMap.set(
-      bucketIndex,
-      Date.now() + this.refreshBucketTimerDefault,
-    );
-    // If the bucket was pending a refresh we remove it
-    this.refreshBucketQueueRemove(bucketIndex);
-    if (bucketIndex === this.refreshBucketNext) {
-      // Bucket is same as next bucket, this affects the timer
-      this.refreshBucketRefreshTimer();
+  @ready(new nodesErrors.ErrorNodeManagerNotRunning())
+  public async updateRefreshBucketDelay(
```

Comment:
In the process of refreshing the bucket delay, the `getTask` and `updateTask` cannot share the same transaction context.

```
const task = await getTask();
try {
  await updateTask(task.id);
} catch (e) {
  if (e instanceof ErrorTaskMissing) {
     // create it
  } else if (e instanceof ErrorTaskRunning) {
    // If it is already running
    // Do nothing!
  }
}
```

This is because if the `getTask` and `updateTask` shared a context, even though `updateTask` is locking the `TaskId`, the snapshot is created at `getTask`, and 2 transactions doing a get, then put, can have a conflict. We don't want a conflict here because we don't expect a conflict exception due to task mutation when interacting with the nodes domain.

So here, we are relying on the exceptions of `updateTask` to know whether we can do what we want.

Reply:
This has been fixed.